src/domain/accounts/services/src/access_token_service_impl.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use chrono::Utc;
+use dill::*;
+use kamu_accounts::{
+    AccessToken,
+    AccessTokenPaginationOpts,
+    AccessTokenRepository,
+    AccessTokenService,
+    Account,
+    CreateAccessTokenError,
+    EncodeTokenError,
+    GetAccessTokenError,
+    KamuAccessToken,
+};
+use opendatafabric::{AccountID, Multihash};
+use rand::{self, Rng};
+use uuid::Uuid;
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub const ACCESS_TOKEN_PREFIX: &str = "ka";
+pub const ACCESS_TOKEN_BYTES_LENGTH: usize = 16;
+pub const ENCODED_ACCESS_TOKEN_LENGTH: usize = 61;
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct AccessTokenServiceImpl {
+    access_token_repository: Arc<dyn AccessTokenRepository>,
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+#[scope(Singleton)]
+#[interface(dyn AccessTokenService)]
+impl AccessTokenServiceImpl {
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn new(access_token_repository: Arc<dyn AccessTokenRepository>) -> Self {
+        Self {
+            access_token_repository,
+        }
+    }
+
+    pub fn generate_access_token_impl(&self) -> KamuAccessToken {
+        let mut random_token_bytes = [0_u8; ACCESS_TOKEN_BYTES_LENGTH];
+        rand::thread_rng().fill(&mut random_token_bytes);
+
+        let token_checksum = crc32fast::hash(&random_token_bytes);
+        let token_id = Uuid::new_v4();
+        let full_token_bytes = [
+            token_id.as_bytes(),
+            random_token_bytes.as_slice(),
+            token_checksum.to_be_bytes().as_slice(),
+        ]
+        .concat();
+        let base32_token = base32::encode(base32::Alphabet::Crockford, &full_token_bytes);
+
+        KamuAccessToken {
+            prefix: ACCESS_TOKEN_PREFIX.to_string(),
+            id: token_id,
+            random_bytes: random_token_bytes,
+            random_bytes_hash: Multihash::from_digest_sha3_256(&random_token_bytes),
+            checksum: token_checksum,
+            composed_token: format!("{ACCESS_TOKEN_PREFIX}_{}", &base32_token),
+            base32_token,
+        }
+    }
```

Comment:
This operation does not use `self`, so it's misplaced. It could be simply a part of `KamuAccessToken::new()`